### PR TITLE
Making NO_INTERVALS to be used by clients of Lucene

### PR DESCRIPTION
**Pull request by expani**
### Description
A [bug](https://github.com/opensearch-project/OpenSearch/issues/13616) was reported in OpenSearch which caused Interval Queries containing sub-query in rules to fail with `Sub-iterators of ConjunctionDISI are not on the same document ` if an no_match query was generated after going through the search analyser. 

A similar issue was found in Lucene a few years back and was fixed by [this PR](https://github.com/apache/lucene/pull/11760) 

I am proposing to make `NO_INTERVALS` present in Lucene as `public` so it can be used by it's clients like OpenSearch rather than creating a clone of the same.

### CODE REVIEW

**Review by jainankitk — APPROVED** [review]

**Comment by romseygeek** [comment]
LGTM.  Can you add an entry to CHANGES.txt?

**Comment by expani** [comment]
@romseygeek Should I add only under `Lucene 10.0.0` or other versions as well ?

**Comment by romseygeek** [comment]
We can backport this to 9x so add the entry under the latest 9x version.

**Comment by expani** [comment]
@jainankitk @romseygeek Requesting an approval.

**Comment by expani** [comment]
Added @romseygeek

File: lucene/CHANGES.txt [hunk 1]
```diff
@@ -362,6 +362,8 @@ Other
 
 * GITHUB#13077: Add public getter for SynonymQuery#field (Andrey Bozhko)
 
+* GITHUB#13385: Make NO_INTERVALS source as public to be used by Lucene clients instead of creating clones themselves.
```
[the remark this answers is not in the document]
Comment (expani) [reply lucene/CHANGES.txt]: Added

**Review by romseygeek — APPROVED** [review]

**Review by uschindler — COMMENTED** [review]
Hi, I have one addition I stumbled over serveral times. There's no "official" way to get an IntervalSource that matches no documents/intervals (I have this sometimes in a query parser for Solr to match complex patent restrictions). This class is pkg private:
https://github.com/expani/lucene/blob/c1c70de765cc7ec67762b40846c7be1e1375eb1c/lucene/queries/src/java/org/apache/lucene/queries/intervals/NoMatchIntervalsSource.java

To get an instance there's a workaround, but its undocumented: `public static IntervalsSource Intervals#atLeast(int minShouldMatch, IntervalsSource... sources)`. You can misuse it by passing 1 and an empty array: `Intervals.atLeast(1)`

I wonder how this differs from this one. The implementation there has no iterator, while this one here has.

I'd suggest to merge those implementations and instead of exposing a constant to end user add another static method like `Intervals#noMatch("reason")`.

Unfortunately this is already merged, but `IntervalBuilder` looks like wrong class for that.

**Comment by uschindler** [comment]
Should I open a new issue, or should we revert this one and give a better solution?

**Comment by romseygeek** [comment]
Let's open a new issue and we can discuss a bit more about when you need such a thing and where it should live.

**Comment by uschindler** [comment]
OK. Just for brevity: How does the one made public here differs from the already existing class (except that one returns null iterator while the other one returns an empty iterator)?

**Comment by uschindler** [comment]
See #13388

**Comment by romseygeek** [comment]
From a quick look we can indeed replace everything with NoMatchIntervalIterator.  I think there are some other places that we can make use of the `null` shortcut as well - for example, DisjunctionIntervalsSource.create() can return `null` if its input collection is empty.

**Comment by uschindler** [comment]
This problem is often seen with patent search. I opened https://github.com/apache/lucene/issues/13388

There are more issues, but making that constant publicly available is wrong. There should be a method to create an empty interval with a reason.

**Pull request merged**